File: src/main/scala/simple/listNode/TwohundredAndThreeRemoveElements.scala
package simple.listNode

import common.ListNodeScala

/**
  *
  * @description: 203. 移除链表元素
  *               删除链表中等于给定值 val 的所有节点。
  * @author: sunhaiting
  * @create: 2020-01-05 11:08
  **/
object TwohundredAndThreeRemoveElements {
  def main(args: Array[String]): Unit = {
    val list1 = new ListNodeScala(1)
    val list2 = new ListNodeScala(2)
    val list3 = new ListNodeScala(3)
    val list4 = new ListNodeScala(4)
    val list5 = new ListNodeScala(5)
    list1.next = list2
    list2.next = list3
    list3.next = list3
    list3.next = list4
    list4.next = list5
    val value = 2
    val node = removeElements2(list1, value)
    println(node)
    println(node.x)
    println(node.next.x)
    println(node.next.next.x)
    println(node.next.next.next.x)
  }

  /**
    * 示例:
    *
    * 输入: 1->2->6->3->4->5->6, val = 6
    * 输出: 1->2->3->4->5
    *
    * @param head
    * @param value
    * @return
    */
  def removeElements(head: ListNodeScala, value: Int): ListNodeScala = {
    //添加虚拟头节点
    var head1 = head
    val newHead = new ListNodeScala(0)
    newHead.next = head1
    head1 = newHead

    var currentNode = head1.next
    var previousNode = head1
    while (currentNode != null) {
      if (currentNode.x == value) {
        previousNode.next = currentNode.next
        currentNode.next = null
        currentNode = previousNode.next
      }
      else {
        currentNode = currentNode.next
        previousNode = previousNode.next
      }
    }
    newHead.next
  }

  def removeElements2(head: ListNodeScala, value: Int): ListNodeScala = {
    //添加虚拟头节点，防止删除的元素是第一个
    val newHead = new ListNodeScala()
    newHead.next = head

    var cur = newHead.next
    var pre = newHead

    println(pre)
    println(newHead)

    println(cur)
    println(head)
    while (cur != null) {
      if (cur.x == value) {
        pre.next = cur.next
      } else {
        pre = pre.next
      }
      cur = cur.next
    }
    newHead.next
  }
}

File: src/main/java/simple/listnode/TwoHundredAndThreeRemoveElements.java
package simple.listnode;

import common.ListNode;

import java.util.ArrayList;

/**
 * @description: Leetcode203. 移除链表元素
 * @author: sunhaiting
 * @create: 2020-04-17 10:53
 **/
public class TwoHundredAndThreeRemoveElements {
    public static void main(String[] args) {

        int[] arr = {1,1, 2, 3, 3, 4, 4};
        ListNode listNode = ListNode.arrayToListNode(arr);
        ListNode listNode1 = removeElements2(listNode, 1);

        ArrayList<Integer> arrayList = ListNode.listNodeToArray(listNode1);
        for (int i : arrayList) {
            System.out.println(i);
        }

    }

    /**
     * 输入: 1->2->6->3->4->5->6, val = 6
     * 输出: 1->2->3->4->5
     *
     * @param head
     * @param k
     * @return
     */
    public static ListNode removeElements(ListNode head, int k) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        ListNode pre = dummy;
        ListNode cur = head;
        while (cur!= null) {
            if (cur.value == k) {
                pre.next = cur.next ;
            } else {
                pre = pre.next;
            }
            cur =cur.next;
        }
        return dummy.next;
    }

    /**
     *
     * @param head
     * @param k
     * @return
     */
    public static ListNode removeElements2(ListNode head, int k) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        ListNode pre = dummy;
        while (pre.next!= null) {
            if (pre.next.value == k) {
                pre.next = pre.next.next ;
            } else {
                pre = pre.next;
            }
        }
        return dummy.next;
    }



}
